# TODO: Reorient input list of seqs

# TODO: Run string decomposer

import pandas as pd


configfile: "config.yaml"


CENS = pd.read_csv(config["contigs"], sep="\t")
MISASSEMBLIES = pd.read_csv(config["misassemblies"], sep="\t")

CHR = CENS["chromosome"].unique()
ORT = CENS["orientation"].unique()
GRP = CENS["group"].unique()

print(CENS)
print(MISASSEMBLIES)
print(CHR, ORT, GRP)


# Index fasta if not already done.
rule index_fa:
    input:
        CENS["path"],
    output:
        [f"{fa}.fai" for fa in CENS["path"]],
    shell:
        """
        for fa in {input}; do
            samtools faidx ${{fa}}
        done
        """


rule group_inputs:
    input:
        fai=rules.index_fa.output,
        cens_data=config["contigs"],
    output:
        "",
    shell:
        """
        """


# rule stringdecomposer:
#     input:
#         # chr21
#         monomers="data/monomers/chm13_{chr}_monomers.fa",
#         # HG03710_tig00000001
#         fa="data/hprc/sequence/{contig}.fa",
#     output:
#         "",
#     conda:
#         "../env/stringdecomposer.yaml"
#     benchmark:
#         "benchmarks/stringdecomposer.tsv"
#     log:
#         "logs/stringdecomposer.log",
#     shell:
#         """
#         stringdecomposer \
#         {input.fa} \
#         {input.monomers} \
#         --outfile {output} 2> {log}
#         """
# # TODO: Generate config file.
